src/sage/algebras/quatalg/quaternion_algebra.py

```diff
@@ -3182,6 +3182,172 @@ def multiply_by_conjugate(self, J):
+            sage: I1.pushforward(I2, side="right")
+            Traceback (most recent call last):
+            ...
+            ValueError: self and J must have the same right-orders
+            sage: I1.conjugate().pushforward(I2.conjugate())
+            Fractional ideal (1/2 + 3/2*j + 10*k, 1/10*i + 2*j + 39/10*k, 3*j, 15*k)
+            sage: I1.conjugate().pushforward(I2.conjugate(), side="left")
+            Traceback (most recent call last):
+            ...
+            ValueError: self and J must have the same left-orders
+            sage: I1.pushforward(I1, side="left")
+            Traceback (most recent call last):
+            ...
+            ValueError: self and J must have coprime norms
+            sage: I3 = B.ideal([1/2 + 13/2*j + 6*k, 1/2*i + 3*j + 13/2*k, 9*j, 9*k])
+            sage: I3.pushforward(I3*(1/3), side="left")
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Quaternion ideal pushforward not implemented for non integral ideals
+        """
+        if not isinstance(J, QuaternionFractionalIdeal_rational):
+            raise TypeError("can only pushforward through an ideal")
+
+        if side == "left":
+            if self.left_order() != J.left_order():
+                raise ValueError("self and J must have the same left-orders")
+            if not self.is_integral() or not J.is_integral():
+                raise NotImplementedError("Quaternion ideal pushforward not implemented for non integral ideals")
+            Jnorm = J.norm()
+            if gcd(self.norm(), Jnorm) != 1:
+                raise ValueError("self and J must have coprime norms")
+            I3 = J.conjugate() * self.intersection(J)
+            I3 = I3 * (1 / Jnorm)
+            return I3
+
+        if side == "right":
+            if self.right_order() != J.right_order():
+                raise ValueError("self and J must have the same right-orders")
+            return self.conjugate().pushforward(J.conjugate(), side="left").conjugate()
+
+        if side is None:
+            same_left_order = self.left_order() == J.left_order()
+            same_right_order = self.right_order() == J.right_order()
+            if not same_left_order and not same_right_order:
+                raise ValueError("self and J must share a left or right order")
+            if same_left_order and same_right_order:
+                raise ValueError("self and J have same left and right orders, side of pushforward must be specified")
+            if same_left_order:
+                return self.pushforward(J, side="left")
+            return self.pushforward(J, side="right")
+
+        raise ValueError("side must be left, right or None")
+
+    def pullback(self, J, side=None):
+        """
+        Compute the ideal which is the pullback of ``self`` through an ideal ``J``.
+        Uses Lemma 2.1.7 of [Ler2022]_. Only works for integral ideals.
```

Comment:
```suggestion
    def pullback(self, J, side=None):
        r"""
        Compute the ideal which is the pullback of ``self`` through an ideal ``J``.

        Uses Lemma 2.1.7 of [Ler2022]_. Only works for integral ideals.
```
This method has no examples, only tests. You should have at least one example, if not more.

Reply:
I added the above change and an example for each function in a separate commit. I felt one example was enough seeing as the functions are quite simple, but can add more if you think there is value in that